### PR TITLE
Simplify the logic by removing dead code and enhance logging

### PR DESCRIPTION
**Pull request by bsdnet**
Signed-off-by: Roy Yang <royyang@google.com>

**What type of PR is this?**
 /kind cleanup
/kind deprecation
/kind documentation

**What this PR does / why we need it**:
When debugging sig-node related tests, find a couple of issues on image selection:
1.) More image-config.yaml will use image_family, but this will result to multiple similar logs
W0528 07:28:02.701] I0528 07:28:02.700935    5119 run_remote.go:500] Found image "\"cos-77-12371-114-0\" created \"2019-10-31 10:35:48.326 -0700 -0700\"" based on regex "" and family "cos-77-lts" in project "cos-cloud"
https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/91138/pull-kubernetes-node-e2e-containerd/1265906725848879104/build-log.txt
2.) The follow does not clearly specify which image is picked finally in the test.
I0528 07:28:03.249] Initializing e2e tests using image cos-stable.
I0528 07:28:03.254] Initializing e2e tests using image ubuntu.
3.) PreviousImage is not used at all, remove the logic
https://github.com/bsdnet/kubernetes/blob/56ad0cefbdbd70a495c1e3d548b82d4e194385af/test/e2e_node/runner/remote/run_remote.go#L160
4.) getGCEImage always return one image since previous_image is not used, and always 0. 
https://github.com/bsdnet/kubernetes/blob/56ad0cefbdbd70a495c1e3d548b82d4e194385af/test/e2e_node/runner/remote/run_remote.go#L160

**Which issue(s) this PR fixes**:

**Special notes for your reviewer**:

This patch removes some unused code, enhance the logging, make debugging test failure easier.

**Does this PR introduce a user-facing change?**:
NONE 

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:
NONE

### CODE REVIEW

File: test/e2e_node/runner/remote/run_remote.go
Comment (MHBauer) [inline test/e2e_node/runner/remote/run_remote.go:83]: looks like it would be an improvement!

File: test/e2e_node/runner/remote/run_remote.go
Comment (MHBauer) [inline test/e2e_node/runner/remote/run_remote.go:151]: don't need the yaml part of this comment.

File: test/e2e_node/runner/remote/run_remote.go
Comment (MHBauer) [inline test/e2e_node/runner/remote/run_remote.go:171]: two 'n' in second internal, innternal -> internal.

File: test/e2e_node/runner/remote/run_remote.go
Comment (MHBauer) [inline test/e2e_node/runner/remote/run_remote.go:219]: doesn't have to be yaml, could be json.
"parse image from given config file and ...."

File: test/e2e_node/runner/remote/run_remote.go
Comment (MHBauer) [inline test/e2e_node/runner/remote/run_remote.go:230]: the code says unmarshal and map, comment is not informative.

File: test/e2e_node/runner/remote/run_remote.go
Comment (MHBauer) [inline test/e2e_node/runner/remote/run_remote.go:480]: Are we 100% sure this behavior is not used anywhere?

File: test/e2e_node/runner/remote/run_remote.go
Comment (MHBauer) [inline test/e2e_node/runner/remote/run_remote.go:512]: different way of solving the silent failure issue.

File: test/e2e_node/runner/remote/run_remote.go
Comment (MHBauer) [inline test/e2e_node/runner/remote/run_remote.go:510]: don't need to do the append, can directly return imageObjs[0].name.

File: test/e2e_node/runner/remote/run_remote.go
Comment (MHBauer) [inline test/e2e_node/runner/remote/run_remote.go:323]: should be using klog, not printf

this solves 2.) ?

**Review by MHBauer — COMMENTED** [review]
I'd have to check that the previous images is never used before I'm comfortable with an lgtm.

also going to fail due to the error returned from finding 0 images until we clean up the test-infra image-configs. 

similar to my  #91470

File: test/e2e_node/runner/remote/run_remote.go
Comment (bsdnet) [reply test/e2e_node/runner/remote/run_remote.go:480]: I have searched both test-infra and kubernetes repo, and also chase the code history. Looks it was not used from day 1.

File: test/e2e_node/runner/remote/run_remote.go
Comment (bsdnet) [reply test/e2e_node/runner/remote/run_remote.go:512]: Yes, I think about your change when refactoring the code here. Maybe we should put here since we expect at least one image found. No image means error, what do you think?

File: test/e2e_node/runner/remote/run_remote.go
Comment (bsdnet) [reply test/e2e_node/runner/remote/run_remote.go:510]: Yes, good catch!

File: test/e2e_node/runner/remote/run_remote.go
Comment (bsdnet) [reply test/e2e_node/runner/remote/run_remote.go:323]: Yes, this will give clear idea which image is used.  Output example is:
ubuntu/project/ubuntu-18-04 or
cos-stable/cos-cloud/cos-77-12347-237-0

File: test/e2e_node/runner/remote/run_remote.go
Comment (bsdnet) [reply test/e2e_node/runner/remote/run_remote.go:230]: Updated and make it more meaningful

File: test/e2e_node/runner/remote/run_remote.go
Comment (MHBauer) [reply test/e2e_node/runner/remote/run_remote.go:512]: I agree if nobody is using the function anywhere else. I would need to check again and do another review.

**Comment by matthyx** [comment]
/ok-to-test
Let's see...

**Comment by bsdnet** [comment]
pull-kubernetes-node-e2e failed because this patch used to catch image non-exists failures. #17770 will correct them. Other two looks unrelated.

File: test/e2e_node/runner/remote/run_remote.go
Comment (alejandrox1) [inline test/e2e_node/runner/remote/run_remote.go:229]: nit:
```suggestion
		// Unmarshal then given image config file. All images for this test run will be organized in a map
		// where each GCEImage value will be associated with the image's name as a key.
```

or something like that, just be more explicit about what is going on.

File: test/e2e_node/runner/remote/run_remote.go
Comment (alejandrox1) [inline test/e2e_node/runner/remote/run_remote.go:247]: nit
```suggestion
			// Convert given image into an internalGCEImage.
```

File: test/e2e_node/runner/remote/run_remote.go
Comment (alejandrox1) [inline test/e2e_node/runner/remote/run_remote.go:470]: Since `previousImages` is being removed can we add a comment about the behaviour of this function, something like
```suggestion
// getGCEImages creates a list of image names based on regex and given GCE project.
// It will return the latest image from that list.
```

File: test/e2e_node/runner/remote/run_remote.go
Comment (alejandrox1) [inline test/e2e_node/runner/remote/run_remote.go:471]: Given the changes being made the function signature should also change 
```suggestion
func getGCEImages(imageRegex, imageFamily string, project string) (string, error) {
```

File: test/e2e_node/runner/remote/run_remote.go
Comment (alejandrox1) [inline test/e2e_node/runner/remote/run_remote.go:505]: Since this will always return the latest image, might as well just change this to
```go
return imageObjs[0].name, nil
```

File: test/e2e_node/runner/remote/run_remote.go
Comment (alejandrox1) [inline test/e2e_node/runner/remote/run_remote.go:239]:
```suggestion
image, err = getGCEImages(imageConfig.ImageRegex, imageConfig.ImageFamily, imageConfig.Project)
```

Since we are defaulting to always return the latest image for the given regex and project combo.

File: test/e2e_node/runner/remote/run_remote.go
Comment (alejandrox1) [inline test/e2e_node/runner/remote/run_remote.go:290]: nit:
```suggestion
	// Make sure project is set.
```

**Review by alejandrox1 — COMMENTED** [review]
Left a couple comments but overall looks good!
Thank you @bsdnet 

/release-note-none
/priority important-longterm

**Comment by alejandrox1** [comment]
/retitle Simplify the logic by removing dead code and enhance logging
and checking if test failures are flakes
/retest

**Comment by MHBauer** [comment]
kubernetes/test-infra#17770 should help with the failures.

**Comment by alejandrox1** [comment]
> @alejandrox1: you can not set the release note label to "release-note-none" because the PR has the label "kind/deprecation".
> 
> In response to [this](https://github.com/kubernetes/kubernetes/pull/91543#issuecomment-637039615):
> 
> > /release-note-none
> 
> Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md). If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.

ha :sweat_smile: what do you all think about labeling this with
```
/kind cleanup
/remove-kind deprecation
```
?
Otherwise this won't merge without a release note.

File: test/e2e_node/runner/remote/run_remote.go
Comment (MHBauer) [inline test/e2e_node/runner/remote/run_remote.go:237]: can be single string after change to getGCEImages

File: test/e2e_node/runner/remote/run_remote.go
Comment (MHBauer) [inline test/e2e_node/runner/remote/run_remote.go:248]: as it falls through as a single image, the loop can go away.

File: test/e2e_node/runner/remote/run_remote.go
Comment (MHBauer) [inline test/e2e_node/runner/remote/run_remote.go:215]: if we can only pull one out, we don't need a map.

**Review by MHBauer — COMMENTED** [review]
I agree that PreviousImage is not used. This seems to trickle through deeply into the code and remove many possibilities of loops to have more than one entry.

The closer I look, the more weirdness I find. So I'm stopping before I leave a zillion comments. I think this is an improvement overall.

I know I see multiple images in single runs, but it doesn't seem to be physically possible after this change. is run_remote.go called more than once in a run?

File: test/e2e_node/runner/remote/run_remote.go
Comment (bsdnet) [reply test/e2e_node/runner/remote/run_remote.go:283]: This is coding style and consistent issue.

File: test/e2e_node/runner/remote/run_remote.go
Comment (alejandrox1) [reply test/e2e_node/runner/remote/run_remote.go:290]: I like this idea better. my comment was purely based on semantics but we should definetely address why that step is needed.

File: test/e2e_node/runner/remote/run_remote.go
Comment (bsdnet) [reply test/e2e_node/runner/remote/run_remote.go:170]: Done

File: test/e2e_node/runner/remote/run_remote.go
Comment (bsdnet) [reply test/e2e_node/runner/remote/run_remote.go:215]: I think the scope is keep changing. This is a minor refractor to clean up things. I do not  think we should rewrite existing code, Good for that to be in a separate CL.

**Comment by alejandrox1** [comment]
> I know I see multiple images in single runs, but it doesn't seem to be physically possible after this change. is run_remote.go called more than once in a run?

**tl;dr** a set of e2e tests seems to be executed for each "image" in the node config file, which is different from the change being made here.
For example, if we used https://github.com/kubernetes/test-infra/blob/master/jobs/e2e_node/image-config-serial-cpu-manager.yaml , we would run 2 sets of e2e tests, once in Ubuntu and once is COS-stable.

If we wanted to run different e2e tests on different images we would have to specify a `test` field on the images, such as in 
https://github.com/kubernetes/test-infra/blob/6e231e22b8a44e74f8c62ef4324bdd8ddf0407b3/jobs/e2e_node/benchmark-config.yaml#L8-L9


---

The place where tests are executed is through 
https://github.com/kubernetes/kubernetes/blob/5bcc33ee94caa138b72db09d3783ce63263a77f2/test/e2e_node/runner/remote/run_remote.go#L329-L331

It calls the `testImage` function
https://github.com/kubernetes/kubernetes/blob/5bcc33ee94caa138b72db09d3783ce63263a77f2/test/e2e_node/runner/remote/run_remote.go#L522

which creates a `TestResult` channel by calling the `testHost` function inside of it
https://github.com/kubernetes/kubernetes/blob/5bcc33ee94caa138b72db09d3783ce63263a77f2/test/e2e_node/runner/remote/run_remote.go#L425
This function executes the actual e2e tests by running `k/k/test/e2e_node/remote.RunRemote`
https://github.com/kubernetes/kubernetes/blob/5bcc33ee94caa138b72db09d3783ce63263a77f2/test/e2e_node/runner/remote/run_remote.go#L455

---

So actual e2e test execution is
```
testImage()
    |
    ----> testHost()
                |
                ----> remote.RunRemote()
                            |
                            ----> Execute e2e test in VM
```

The thing to notice is that `testImage` is called in `main` inside of a go routine here
https://github.com/kubernetes/kubernetes/blob/5bcc33ee94caa138b72db09d3783ce63263a77f2/test/e2e_node/runner/remote/run_remote.go#L325-L332
and it is executed once per entry in `gceImages.images`.

`gceImages` is defined here
https://github.com/kubernetes/kubernetes/blob/5bcc33ee94caa138b72db09d3783ce63263a77f2/test/e2e_node/runner/remote/run_remote.go#L218-L220

the image config files,
```yaml
images:
  cosstable2-resource1:
    image: cos-77-12371-274-0
    project: cos-cloud
    machine: n1-standard-1
  cosstable2-resource2:
    image: cos-77-12371-274-0
    project: cos-cloud
    machine: n1-standard-1
...
```

are parsed into `ImageConfig` objects
https://github.com/kubernetes/kubernetes/blob/5bcc33ee94caa138b72db09d3783ce63263a77f2/test/e2e_node/runner/remote/run_remote.go#L232-L233

and then for each key in the config file
https://github.com/kubernetes/kubernetes/blob/5bcc33ee94caa138b72db09d3783ce63263a77f2/test/e2e_node/runner/remote/run_remote.go#L237
i.e., `cosstable2-resource1` and `cosstable2-resource2`, in the example yaml i included above,
and actual gce image is added to the `gceImages` object
https://github.com/kubernetes/kubernetes/blob/5bcc33ee94caa138b72db09d3783ce63263a77f2/test/e2e_node/runner/remote/run_remote.go#L271

So for each "image" in the node config, a set of e2e tests will still be executed.

File: test/e2e_node/runner/remote/run_remote.go
Comment (alejandrox1) [reply test/e2e_node/runner/remote/run_remote.go:215]: This one is separate from the change to `getGCEImages`.
`getGCEImages` only gets one image, but a node config can have multiple images, see https://github.com/kubernetes/kubernetes/pull/91543#issuecomment-637141822

Comment (alejandrox1) [reply test/e2e_node/runner/remote/run_remote.go:215]: Which I think means that this part should stay as is.

**Comment by MHBauer** [comment]
@alejandrox1  you get all the points today. I checked it out and saw the scope change.

File: test/e2e_node/runner/remote/run_remote.go
Comment (MHBauer) [reply test/e2e_node/runner/remote/run_remote.go:215]: yup, I read it wrong here, scope changed.

Comment (bsdnet) [reply test/e2e_node/runner/remote/run_remote.go:215]: yes, alejandrox1@ is right.

File: test/e2e_node/runner/remote/run_remote.go
Comment (MHBauer) [inline test/e2e_node/runner/remote/run_remote.go:230]: super!

File: test/e2e_node/runner/remote/run_remote.go
Comment (MHBauer) [inline test/e2e_node/runner/remote/run_remote.go:321]: so at the point this is printed, we've already made the choice of which image to run, and are getting ready to run it, awesome!

**Review by MHBauer — APPROVED** [review]
/lgtm
I'm happy with it.

File: test/e2e_node/runner/remote/run_remote.go
Comment (bsdnet) [reply test/e2e_node/runner/remote/run_remote.go:321]: yes, images and software version are important for troubleshooting. We should print as detailed as possible.

**Comment by bsdnet** [comment]
/test pull-kubernetes-verify

./hack/update-gofmt.sh has not any issue.

File: test/e2e_node/runner/remote/run_remote.go
Comment (cynepco3hahue) [inline test/e2e_node/runner/remote/run_remote.go:470]: IIUC regular expression can be used to get a number of images and not the only one image, why do we remove this functionality?

File: test/e2e_node/runner/remote/run_remote.go
Comment (cynepco3hahue) [inline test/e2e_node/runner/remote/run_remote.go:500]: Again the same question, why do we want to return only the first image?

Comment (bsdnet) [reply test/e2e_node/runner/remote/run_remote.go:500]: We want to use the latest image after sorting.

File: test/e2e_node/runner/remote/run_remote.go
Comment (bsdnet) [reply test/e2e_node/runner/remote/run_remote.go:470]: Though we get multiple images, but the function always return one from same project or image family. Same as your below question. Regex is designed to match all available images to avoid yaml updates. That should be the design purpose.

Comment (cynepco3hahue) [reply test/e2e_node/runner/remote/run_remote.go:470]: I see now, because of `gceImages.images[shortName] = gceImage` we anyway use only one image, thank for clarification.

**Comment by bsdnet** [comment]
/assign @derekwaynecarr @matthyx  @klueska 

Let me know what needs to be done for this to be merged

**Comment by matthyx** [comment]
One of the approvers need to approve...
/assign @derekwaynecarr

**Pull request merged**
